File: comportamiento/observer.py
from abc import ABC, abstractmethod

class Observer(ABC):
    @abstractmethod
    def update(self, serie, episode):
        pass

class Serie:
    def __init__(self, name):
        self.name = name
        self._observers = []

    def add_observer(self, observer):
        self._observers.append(observer)

    def remove_observer(self, observer):
        self._observers.remove(observer)

    def notify_observers(self, episode):
        for observer in self._observers:
            observer.update(self, episode)

    def new_episode(self, episode):
        print(f"¡Nuevo episodio de {self.name}: {episode}!")
        self.notify_observers(episode)

class Usuario(Observer):
    def __init__(self, name):
        self.name = name

    def update(self, serie, episode):
        print(f"{self.name}: ¡Nuevo episodio de {serie.name} disponible! ({episode})")

class PlataformaStreaming:
    def __init__(self):
        self.series = {}

    def add_serie(self, serie):
        self.series[serie.name] = serie

    def release_episode(self, serie_name, episode):
        if serie_name in self.series:
            self.series[serie_name].new_episode(episode)
        else:
            print(f"Error: La serie '{serie_name}' no existe en la plataforma.")

if __name__ == "__main__":
    plataforma = PlataformaStreaming()

    serie1 = Serie("Breaking Bad")
    serie2 = Serie("Game of Thrones")

    plataforma.add_serie(serie1)
    plataforma.add_serie(serie2)

    usuario1 = Usuario("Alice")
    usuario2 = Usuario("Bob")
    usuario3 = Usuario("Charlie")

    serie1.add_observer(usuario1)
    serie1.add_observer(usuario2)
    serie2.add_observer(usuario3)

    print("=== Lanzamiento de nuevos episodios ===")
    plataforma.release_episode("Breaking Bad", "Episodio 5x01")
    plataforma.release_episode("Game of Thrones", "Episodio 8x03")
    plataforma.release_episode("Stranger Things", "Episodio 4x01")  

File: comportamiento/observer.java
import java.util.*;

interface Observer {
    void update(Serie serie, String episode);
}

class Serie {
    private String name;
    private List<Observer> observers;

    public Serie(String name) {
        this.name = name;
        this.observers = new ArrayList<>();
    }

    public String getName() {
        return name;
    }

    public void addObserver(Observer observer) {
        observers.add(observer);
    }

    public void removeObserver(Observer observer) {
        observers.remove(observer);
    }

    public void notifyObservers(String episode) {
        for (Observer observer : observers) {
            observer.update(this, episode);
        }
    }

    public void newEpisode(String episode) {
        System.out.println("¡Nuevo episodio de " + name + ": " + episode + "!");
        notifyObservers(episode);
    }
}

class Usuario implements Observer {
    private String name;

    public Usuario(String name) {
        this.name = name;
    }

    @Override
    public void update(Serie serie, String episode) {
        System.out.println(name + ": ¡Nuevo episodio de " + serie.getName() + " disponible! (" + episode + ")");
    }
}

class PlataformaStreaming {
    private Map<String, Serie> series;

    public PlataformaStreaming() {
        this.series = new HashMap<>();
    }

    public void addSerie(Serie serie) {
        series.put(serie.getName(), serie);
    }

    public void releaseEpisode(String serieName, String episode) {
        if (series.containsKey(serieName)) {
            series.get(serieName).newEpisode(episode);
        } else {
            System.out.println("Error: La serie '" + serieName + "' no existe en la plataforma.");
        }
    }
}

public class Main {
    public static void main(String[] args) {
        PlataformaStreaming plataforma = new PlataformaStreaming();

        Serie serie1 = new Serie("Breaking Bad");
        Serie serie2 = new Serie("Game of Thrones");

        plataforma.addSerie(serie1);
        plataforma.addSerie(serie2);

        Usuario usuario1 = new Usuario("Alice");
        Usuario usuario2 = new Usuario("Bob");
        Usuario usuario3 = new Usuario("Charlie");

        serie1.addObserver(usuario1);
        serie1.addObserver(usuario2);
        serie2.addObserver(usuario3);

        System.out.println("=== Lanzamiento de nuevos episodios ===");
        plataforma.releaseEpisode("Breaking Bad", "Episodio 5x01");
        plataforma.releaseEpisode("Game of Thrones", "Episodio 8x03");
        plataforma.releaseEpisode("Stranger Things", "Episodio 4x01");  
    }
}
